marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java

```diff
@@ -29,96 +29,98 @@
-		  }
-		  authorizationTokenValue = expiringSAMLAuth.getAuthorizationToken();
-		  
-		  
-		  if(isCallbackExecuting.compareAndSet(false, true))
-		      threshold = expiry.getEpochSecond();
-		  
-	  }
-	  
-	  private static void checkAuthorizationExpiry(Instant instant) {
-	      if(instant == null) {
-              throw new IllegalArgumentException("SAML authentication does not have expiry value.");
-          }
-          if(instant.isBefore(Instant.now())) {
-              throw new IllegalArgumentException("SAML authentication token has expired.");
-          }
-	  }
-	  static void renew(ExpiringSAMLAuth expiringSamlAuth) {
-	      try {
-	      Instant newInstant = renewer.apply(expiringSamlAuth);
-	      
-	      checkAuthorizationExpiry(newInstant);
-	      long current = Instant.now().getEpochSecond();
-	      threshold = current+ ((newInstant.getEpochSecond() - current)/2);
-	      } finally {
-            isCallbackExecuting = new AtomicBoolean(false);
-        }
-	  }
     @Override
-    public void run() {
-        renew(expiringSAMLAuth);
-        
+    public Response intercept(Chain chain) throws IOException {
+        Request request = chain.request();
+        if (authorizer != null) {
+            if ((expiringSAMLAuth == null || threshold <= Instant.now().getEpochSecond())
+                    && isCallbackExecuting.compareAndSet(false, true)) {
+                authorizeCallback();
+            }
+        } else if (renewer != null) {
+            if (threshold <= Instant.now().getEpochSecond()) {
+                RenewExpiry renewExpiry = new RenewExpiry();
+                if (isCallbackExecuting.compareAndSet(false, true))
+                    Executors.defaultThreadFactory().newThread(renewExpiry).start();
+
+            }
+        }
+        String samlHeaderValue = RESTServices.AUTHORIZATION_TYPE_SAML + " " + RESTServices.AUTHORIZATION_PARAM_TOKEN
+                + "=" + authorizationTokenValue;
+        Request authenticatedRequest = request.newBuilder().header(RESTServices.HEADER_AUTHORIZATION, samlHeaderValue)
+                .build();
+        return chain.proceed(authenticatedRequest);
+    }
+
+    private void authorizeCallback() {
+        try {
+            expiringSAMLAuth = authorizer.apply(expiringSAMLAuth);
+
+            if (expiringSAMLAuth == null) {
+                throw new IllegalArgumentException("SAML Authentication cannot be null");
+            }
+
+            if (expiringSAMLAuth.getAuthorizationToken() == null) {
+                throw new IllegalArgumentException("SAML Authentication token cannot be null");
+            }
+            authorizationTokenValue = expiringSAMLAuth.getAuthorizationToken();
+            checkAuthorizationExpiry(expiringSAMLAuth.getExpiry());
+        } finally {
+            isCallbackExecuting.set(false);
+        }
+
+    }
+
+    private void checkAuthorizationExpiry(Instant instant) {
```

Comment:
Would the name setThreshold(Instant expiry) better convey what the method is doing?